Sources/NIOSSL/SSLContext.swift

```diff
@@ -520,6 +527,83 @@ extension NIOSSLContext {
     }
 }
 
+// For utilizing SNI callbacks from a SSLContext
+extension NIOSSLContext {
+    
+    private static func setSniCallback(context: OpaquePointer) throws {
+        
+        CNIOBoringSSL_SSL_CTX_set_tlsext_servername_callback(context) { (ssl, adPointer, argPointer) in
+            guard let ssl = ssl else {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            // Make sure we get the server name extracted.
+            guard let servername = CNIOBoringSSL_SSL_get_servername(ssl, CNIOBoringSSL.TLSEXT_NAMETYPE_host_name) else
+            {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+            
+            let serverNameString = String(cString: servername)
+            
+            // Can do nothing about empty indicators.
+            if (serverNameString.isEmpty)
+            {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+            
+            // We want to take the SSL pointer and extract the parent Swift object. These force-unwraps are for
+            // safety: a correct NIO program can never fail to set these pointers, and if it does failing loudly is
+            // more useful than failing quietly.
+            let parentCtx = CNIOBoringSSL_SSL_get_SSL_CTX(ssl)!
+            let parentPtr = CNIOBoringSSLShims_SSL_CTX_get_app_data(parentCtx)!
+            let parentSwiftContext: NIOSSLContext = Unmanaged.fromOpaque(parentPtr).takeUnretainedValue()
+
+            // Similarly, this force-unwrap is safe because a correct NIO program can never fail to unwrap this entry
+            // either.
+            guard let userChosenContext = parentSwiftContext.sniManager!.sniInicated(hostname: serverNameString, context: parentSwiftContext) else
+            {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+            
+            // The user may or may not retain a reference to the NIOSSLContext
+            // that they return here, if they provide one. The NIOSSLContext finalizer
+            // explicitly decrements the reference count on the underlying SSL_CTX.
+            //
+            // However, SSL_set_SSL_CTX explicitly increases the reference count and
+            // will decrement the count when the SSL* object is destroyed, thus finally
+            // freeing the raw SSL_CTX object, even if the swift reference is orphaned.
+            //
+            // As such, no memory should be leaked or freed too early in the event that
+            // the user fails to retain their created and return NIOSSLContext. In the
+            // event that they do retain it, the reference count will remain at at-least
+            // 1 and thus they can re-use the same context for multiple connections.
+            
+            guard let nativeReturnVal = CNIOBoringSSL_SSL_set_SSL_CTX(ssl, userChosenContext.sslContext) else
```

Comment:
The above comment is great, but the code as written is not quite right in the face of refactors. If we stopped using `userChosenContext` below, then it would be allowable for Swift to free the `userChosenContext` after extracting the `sslContext` pointer from it, causing `deinit` to run between the `get` and the call to `SSL_set_SSL_CTX`.

To that end, we should wrap this in a call to `withExtendedLifetime` to make sure things behave properly:

```swift
let nativeReturnVal = withExtendedLifetime(userChosenContext) {
    CNIOBoringSSL_SSL_set_SSL_CTX(ssl, userChosenContext.sslContext)
}
```

That way refactors will never affect the correctness of this code.

Reply:
I'll take a look at this. I've never used swift before and I'm trying to imagine how the destructor could be invoked in the same line the object still exists and is being accessed. We don't care at all to retain `nativeReturnResult` and SSL_set_SSL_CTX is a blocking call that immediately runs an atomic increment of the reference count stored in the native object, which will prevent its freeing even if the `userChosenContext` `deinit` is executed.

Reply:
The way to envision it is this: Swift treats all `get` accessors _as if_ they performed a shallow copy of the element. Essentially, `var tempBar = foo.bar` is conceptually equivalent to `var tempBar = foo.getBar()`. Consider, then, this code: `someFunction(foo.bar)`. This is conceptually equivalent to:

```swift
var temporaryBar = foo.getBar()
someFunction(temporaryBar)
```

Here it should be clear why there might be a risk: after `getBar` completes there is no longer a reference to `foo`, and so we can deinit it. This is not a hypothetical: Swift has over time made the ARC checker increasingly aggressive, and I have personally observed it deinit an object in this way.

Reply:
Got it. Thanks.